src/common/config.h

```diff
@@ -2,84 +2,161 @@
-    if (fi_.fail()) {
-      LOG(FATAL) << "cannot open file " << cfgFileName;
-    }
+  explicit ConfigParser(const std::string& path)
+    : line_comment_regex_("^#"),
+      key_regex_(R"rx(^([^#"'=\r\n\t ]+)[\t ]*=)rx"),
+      key_regex_escaped_(R"rx(^(["'])([^"'=\r\n]+)\1[\t ]*=)rx"),
+      value_regex_(R"rx(^([^#"'=\r\n\t ]+)[\t ]*(?:#.*){0,1}$)rx"),
+      value_regex_escaped_(R"rx(^(["'])([^"'=\r\n]+)\1[\t ]*(?:#.*){0,1}$)rx") {
+    NormalizeEOL(path);
+    fi_.open(path);
+    CHECK(!fi_.fail()) << "Cannot open file " << path;
   }
 
   /*!
-   * \brief parse the configure file
+   * \brief Parse configuration file into key-value pairs.
+   * \param path path to configuration file
+   * \return list of key-value pairs
    */
-  std::vector<std::pair<std::string, std::string> > Parse() {
-    std::vector<std::pair<std::string, std::string> > results{};
+  std::vector<std::pair<std::string, std::string>> Parse() {
+    std::vector<std::pair<std::string, std::string>> results;
     char delimiter = '=';
     char comment = '#';
-    std::string line{};
-    std::string name{};
-    std::string value{};
-
-    while (!fi_.eof()) {
-      std::getline(fi_, line);  // read a line of configure file
-      line = line.substr(0, line.find(comment));  // anything beyond # is comment
-      size_t delimiterPos = line.find(delimiter);  // find the = sign
-      name = line.substr(0, delimiterPos);  // anything before = is the name
-      // after this = is the value
-      value = line.substr(delimiterPos + 1, line.length() - delimiterPos - 1);
-
-      if (line.empty() || name.empty() || value.empty())
-        continue;  // skip a line if # at beginning or there is no value or no name.
-      CleanString(&name);  // clean the string
-      CleanString(&value);
-      results.emplace_back(name, value);
+    std::string line;
+    std::string key, value;
+    // Loop over every line of the configuration file
+    while (std::getline(fi_, line)) {
+      if (ParseKeyValuePair(line, &key, &value)) {
+        results.emplace_back(key, value);
+      }
     }
     return results;
   }
 
-  ~ConfigParse() {
-    fi_.close();
-  }
-
  private:
   std::ifstream fi_;
-  std::string allowableChar_ =
-      "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_-./\\";
+  const std::regex line_comment_regex_, key_regex_, key_regex_escaped_,
+    value_regex_, value_regex_escaped_;
+
+ public:
+  /*!
+   * \brief Normalize end-of-line in a file so that it uses LF for all line
+   *        endings.
+   * This is needed because some OSes use CR or CR LF instead.
+   * So we replace all CR with LF. Replacement will be performed in-place.
+   * \param path path to configuration file
+   */
+  static void NormalizeEOL(const std::string& path) {
+    std::FILE* fp = std::fopen(path.c_str(), "rb+");
+    CHECK(fp) << "Cannot open file " << path;
+    int ch;
+    while ((ch = std::fgetc(fp)) != EOF) {
+      if (ch == static_cast<int>('\r')) {
```

Comment:
would this be corrupting user-data, if 'cr' happens to be in it? perhaps, a pathological scenario. is it safe to see if this happens only at the end of line?

Reply:
In the old parser, `\r` was assumed to be a newline character, so we are preserving its behavior.

Reply:
And if we assume that the text file is encoded in UTF-8, then `\r` will never show up as part of a multi-type character.